feat: search results pagination

mohannadzidan opened the pull request:
This PR will add pagination buttons to search results to show more search results 

closes #357, and closes #292

zamitto commented:
Please, always add a screenshot when your PR is releated to UI stuff.

mohannadzidan commented:
![image](https://github.com/hydralauncher/hydra/assets/26430994/efdbb54a-d1d7-4ac5-b7b0-8bd928ec595d)

piradata commented:
could you please add the translation for `en` at least? the others can be added later

piradata commented:
![image](https://github.com/hydralauncher/hydra/assets/16529503/6eb60a95-b247-4474-b870-3f67109562ac)
@mohannadzidan u can use it actually

mohannadzidan commented:
I have updated the code to use `next_page`, `previous_page` keys from the catalogue namespace

![image](https://github.com/hydralauncher/hydra/assets/26430994/e5b2bc20-2d4a-4530-912f-b278e4a3ee5a)

mohannadzidan commented:
@zamitto hey, this feature is ready to merge, can you please review the code?

zamitto commented on src/renderer/src/pages/home/search-results.tsx:
```diff
@@ -50,10 +55,40 @@ export function SearchResults() {
   return (
     <SkeletonTheme baseColor={vars.color.background} highlightColor="#444">
+      <section
+        style={{
+          padding: `${SPACING_UNIT * 3}px ${SPACING_UNIT * 4}px`,
+          display: "flex",
+          width: "100%",
+          justifyContent: "space-between",
+          alignItems: "center",
+          borderBottom: `1px solid ${vars.color.border}`,
+        }}
```
you could move this style object to home.css.ts file

zamitto commented on src/renderer/src/pages/home/search-results.tsx:
```diff
@@ -50,10 +55,40 @@ export function SearchResults() {
+          {t("previous_page", { ns: "catalogue" })}
+        </Button>
+
+        <Button onClick={handleNextPage} theme="outline" disabled={isLoading}>
+          {t("next_page", { ns: "catalogue" })}
+          <ArrowRightIcon />
+        </Button>
```
this button will never get disabled? even if the page has only one result?

piradata commented:
hey @mohannadzidan , need some help here?

mohannadzidan commented:
Hello @piradata and @zamitto

 I've implemented the changes requested in the previous review:

1. Moved pagination inlined styles to home.css for better separation of concerns.
2.  Updated the searchGames function signature for improved clarity and reusability.
3. Added functionality to disable the "next" button when there are no more results to prevent unexpected behavior.

Please take another look at the code and let me know if everything looks good or if you have any further suggestions. Thanks!

piradata commented on src/renderer/src/pages/home/search-results.tsx:
```diff
@@ -16,6 +20,8 @@ import { useNavigate, useSearchParams } from "react-router-dom";
 import * as styles from "./home.css";
 import { buildGameDetailsPath } from "@renderer/helpers";
 
+const PAGE_SIZE = 12;
```
why set the pagesize here if you already set the default on the function itself to 12?

mohannadzidan replied:
I made it like this to increase the code readability and avoid hard coding magic numbers in the code while calculating whether this is the last results page or not

For example if a new contributor saw this line
```ts
setIsEnd(results.length < 12)
```
He will be confused and wonder what does this number  12  mean and why we put it here, right? But by declaring this constant and use it properly in the query it is clearer and will make since to him

piradata replied:
but you can fet this number from the search params the same way you does to get the page no?

piradata commented on src/main/events/catalogue/search-games.ts:
```diff
@@ -1,12 +1,19 @@
+
+const DEFAULT_SEARCH_ARGS: PaginationArgs = {
+  take: 12,
+  skip: 0,
```
instead of skip, you could also just pass the page as a default for zero, and calculate the skip inside the function itself, then you dont need to know the default value for take when you just wannt to go to the page like 5 using the default "take" parameter

mohannadzidan replied:
Actually your suggestion was exactly the previous version of the function when i modified it initially for this pagination feature, but I thought that it is more reusable like this? Like if in the future we had to use this function to display different number of results we will able to do this without revising the code again and modify it

piradata replied:
+/-, like, how it is now, you need to know the `take` parameter to send the `skip` correctly considering the number of the `page` you want, so you basically dont make use of the default value for `take`, as you need to consider it anyway to send the `skip`.

if you send only `take` and `page`, both optional to 12 an 0 respectively, you dont need to care about any of this logic, and can even just send the `page` arg for any request. and you also has access to the `take` (default or not) used on the query to see if you are on the last page (query could also return a parameter with the number of pages tho)

what do you think @zamitto?

henriqueleite42 replied:
I kinda agree with @piradata on this one, but the default to 12 on `take` I think that should be a global default value, indeed of having one here, another one in `src/renderer/src/pages/home/search-results.tsx` and so on.

piradata commented:
@zamitto what do you think on this one?

piradata commented:
@zamitto is this one stil relevant?

zamitto commented:
> @zamitto is this one stil relevant?
Not anymore, we moved the search to our api
We might eventually add pagination to it
But, as now the results are much more relevant and gives what the user is search for, pagination is not a top priority right now
